driver-core/src/main/java/com/datastax/driver/core/Connection.java

```diff
@@ -492,23 +492,51 @@ void setKeyspace(String keyspace) throws ConnectionException {
             return thisFuture;
         }
 
-        logger.trace("{} Setting keyspace {}", this, keyspace);
-        // Note: we quote the keyspace below, because the name is the one coming from Cassandra, so it's in the right case already
-        Future future = write(new Requests.Query("USE \"" + keyspace + '"'));
-        return Futures.transform(future, new Function<Message.Response, Connection>() {
-            @Override
-            public Connection apply(Message.Response response) {
-                if (response instanceof SetKeyspace) {
-                    Connection.this.keyspace = ((SetKeyspace) response).keyspace;
-                    return Connection.this;
-                } else if (response.type == ERROR) {
-                    Responses.Error error = (Responses.Error) response;
-                    throw defunct(error.asException(address));
-                } else {
-                    throw defunct(new DriverInternalError("Unexpected response while setting keyspace: " + response));
+        SetKeyspaceAttempt attempt = new SetKeyspaceAttempt(keyspace);
+
+        // Check for an existing keyspace attempt.
+        while (true) {
+            SetKeyspaceAttempt existingAttempt = targetKeyspace.get();
+            // if existing attempts' keyspace matches what we are trying to set, use it.
+            if (attempt.equals(existingAttempt)) {
+                attempt = existingAttempt;
+                break;
+            } else if (targetKeyspace.compareAndSet(existingAttempt, attempt)) {
+                // Otherwise, start a new keyspace attempt replacing the existing one.
+                // This is ok, as any observers of the existing attempt will still complete.
+                if (existingAttempt != null && !existingAttempt.future.isDone()) {
+                    logger.debug("{} There is already an in progress attempt to set keyspace on this connection for keyspace '{}', beginning set keyspace for '{}'",
+                            this, existingAttempt.keyspace, keyspace);
```

Comment:
I agree we should change this to WARN, and add an explanation like "This can happen if you constantly change the keyspace at runtime". Otherwise the queries fail with errors that are not necessarily easy to debug, like "unconfigured table xxx".
The downside is that we will flood the logs if the client is really degenerate, but that's better than letting all queries fail without explanation.